p32/src/shapes.rs

```diff
@@ -0,0 +1,191 @@
+pub trait Shape: std::fmt::Debug {
+    const NAME: &'static str;
+
+    fn area(&self) -> f32;
+    fn perimeter(&self) -> f32;
+    fn scale(&mut self, factor: f32);
+
+    fn area_to_perimeter(&self) -> f32 {
+        match self.perimeter() {
+            0.0 => 0.0,
+            _ => self.area() / self.perimeter(),
```

Comment:
This method should not special-case the zero case. Instead, `Point` should overwrite default impl for this method.